build(deps): bump libvterm to 0.3-RC1

cryptomilk opened the pull request:
This is a minimal patch to get neovim working with libvterm 0.2 and keep backward compatibility.

cryptomilk committed:
feat(term): add support for libvterm >= 0.2

Signed-off-by: Andreas Schneider <asn@cryptomilk.org>

cryptomilk committed:
build(deps): use libvterm 0.3-rc1

cryptomilk commented:
Looks like there are some unexpected changes with libvterm 0.2. However I don't understand what those tests. Looks like something is shifted by one line sometimes.

clason commented:
Yes, that's what I was getting at. See the comments at the other PR, especially https://github.com/neovim/neovim/pull/16219#issuecomment-958969904.

cryptomilk commented:
I've implemented the buffering but this doesn't fix it.

cryptomilk committed:
refactor(term): use const for title and add len to buf_set_term_title()

cryptomilk committed:
fix(terminal): adopt altscreen test for libvterm 0.2 changes

Due to a rewrite of the resize handling logic in libvterm, the behavior
changed. It's actually fixing a bug.

https://github.com/cryptomilk/libvterm/commit/16b857457575c3fd6ffdb0866d7a190b69c28312

cryptomilk commented:
There is a bug in libvterm 0.2, I have a fix for it: https://github.com/cryptomilk/libvterm/commit/cca90c80799b91203d28176cbf6f62c3e5969e7b

clason commented:
@leonerd

cryptomilk commented:
This looks promising. The macos-11.0 failure seems unrelated as it fails in tree-sitter ...

cryptomilk commented:
@clason Who could help me to understand that failing buffer_spec test better? We need to write a [libvterm test](https://github.com/cryptomilk/libvterm/blob/main/t/63screen_resize.test) for this.

cryptomilk commented:
I have a vterm test which reproduces the issue:

```
INIT
WANTSTATE
WANTSCREEN

!neovim buffer_spec_test handles loss of focus gracefully
RESET
RESIZE 7,50
PUSH "tty ready\r\n"
  ?screen_row 0 = "tty ready"
  ?screen_row 1 = ""
  ?cursor = 1,0
PUSH "rows: 2, cols: 50\r\n"
  ?screen_row 0 = "tty ready"
  ?screen_row 1 = "rows: 2, cols: 50"
  ?cursor = 2,0
RESIZE 2,50
  ?screen_row 0 = "rows: 2, cols: 50"
  ?cursor = 1,0
```

justinmk commented:
This sounds awesome. Please include a brief explanation of the reflow patch in the commit message (and PR). Why isn't it upstreamed? Are we going to have a merge issue later if some form of it gets upstreamed?

cryptomilk commented:
The reflow patch is from the reflow repo which is a rewrite of screen reflow. I've added it to be able that more people can test it.

http://bazaar.leonerd.org.uk/c/libvterm.WIP-REFLOW/

clason commented:
@justinmk this is the offshoot of a discussion in chat between @cryptomilk, @leonerd, and myself. This is an experimental branch to allow testing the (unstable, experimental) reflow branch of libvterm so we can help (by generating failing test cases) to make it stable and turn it into libvterm 0.3. 

This is _not_ meant for merging!

cryptomilk commented:
One test which segfaults does the following:

```
RESET
RESIZE 7,50
```

Create a nvim window with 7 rows, 50 colums:

Then it uses lua to create the following buffer content:

```
1 a
2  a
3   a
4    a
5     a
6      a
7       a
8        a
9         a
10         a
11          a
12           a
```

After creating the buffer the display looks like this (7 rows, ^ indicates the 
curser position):

```
1 ^a
2  a
3   a
4    a
5     a
6      a
(blank line = status bar)
```

then it feeds a 'G' meaning scroll to the end and it expects

```
7       a
8        a
9         a
10         a
11          a
12           ^a
(blank line = status bar)
```

I've tried to create a libvterm test out of this, but the testsuite is completely undocumented.

t/92neovim.test
```
INIT
WANTSTATE
WANTSCREEN

!neovim buffer_spec_test handles loss of focus gracefully
RESET
RESIZE 7,50
PUSH "tty ready\r\n"
  ?screen_row 0 = "tty ready"
  ?screen_row 1 = ""
  ?cursor = 1,0
PUSH "rows: 2, cols: 50\r\n"
  ?screen_row 0 = "tty ready"
  ?screen_row 1 = "rows: 2, cols: 50"
  ?cursor = 2,0
RESIZE 2,50
  ?screen_row 0 = "rows: 2, cols: 50"
  ?cursor = 1,0

!neovim scrollback_spec - does not crash after setting 'number' #14891
RESET
RESIZE 7,50
PUSH "a\r\n"
PUSH " a\r\n"
PUSH "  a\r\n"
PUSH "   a\r\n"
PUSH "    a\r\n"
PUSH "     a\r\n"
PUSH "      a\r\n"
PUSH "       a\r\n"
PUSH "        a\r\n"
PUSH "         a\r\n"
PUSH "          a\r\n"
PUSH "           a"
  ?cursor = 6,12
  scrollrect 0..7,0..50 => -11,+0
  ?cursor = 0,0
```

PUSH will move the cursor. So I tried to scroll back 0,0 but it doesn't work.

```
make && t/run-test.pl t/92neovim.test 
make: Nothing to be done for 'all'.
> neovim buffer_spec_test handles loss of focus gracefully
> neovim scrollback_spec - does not crash after setting 'number' #14891
# line 37: Assert cursor failed:
# Expected: 0,0
# Actual:   6,12
```

I've also sent a mail to @leonerd but he didn't respond. He asks for help, but if you want to provide help and extend the tests he doesn't answer.

leonerd commented:
Hi I'm here. I couldn't see a question about it. It looked like a long list of statements. I'm not sure what to do with these facts.

leonerd commented:
> I've tried to create a libvterm test out of this, but the testsuite is completely undocumented.
Brief 30second summary:
 * `SHOUTY` things in column 1 are commands to tell the test harness to do to libvterm - usually `PUSH` to send more bytes in, but sometimes others. 
 * `lowercase` things are expectations of events that it sends back (e.g. `scrollrect` happened).
 * `?queries` are question/answer assertions - does this fact hold?

Your final test there looks very strange. You query to find that the cursor really is on line 6 - as I'd expect for a 7-line buffer having pushed more than 7 lines of content. It'll sit on the bottom row. As a result of that, a scrollrect event happened. You then query again to see if the cursor is at 0,0 - well that's going to fail because the cursor hasn't moved. You haven't *done* anything since the last query, so it'll be in the same place.

cryptomilk commented:
How do I scroll up to the first line?
How do I scroll down the last line?

cryptomilk commented:
The neovim test fills a buffer, and then scrolls down to the last line and this seems to crash neovim with the libvterm reflow branch. So I'm trying to reproduce what neovim does with the libvterm testsuite.

leonerd commented:
> How do I scroll up to the first line? How do I scroll down the last line?

"scrolling" isn't a concept within libvterm (currently). At the moment, any of that backbuffer behaviour in e.g. pangoterm or neovim's use of `:term` is done entirely by those embedding programs, not libvterm itself. Once content falls off the top of the actual screen (i.e. line 0) that's it. It's gone. The application needs to remember it if it ever wants it back. That's the point of the `sb_pushline` event in fact - it tells the embedding "you're about to lose this line; keep it if you want it"

sassanh commented:
This comment regarding the OSC messages fallback callback may be relevant: https://github.com/neovim/neovim/issues/4349#issuecomment-1237579330

cryptomilk commented:
Updated to latest reflow changes ...

clason commented:
Patching fails on Windows, otherwise looking good \o/

I think we're OK with bumping the minimal required version of libvterm to 0.2 now, so we don't need the `#ifdef` guard (@jamessan ?)

zeertzjq commented on src/nvim/terminal.c:
```diff
@@ -858,13 +863,13 @@ static int term_movecursor(VTermPos new, VTermPos old, int visible, void *data)
   dict_set_var(buf->b_vars,
                STATIC_CSTR_AS_STRING("term_title"),
-               STRING_OBJ(cstr_as_string(title)),
+               STRING_OBJ(((String){ .data = (char *)title, .size = len })),
```
redundant parentheses

leonerd commented on cmake.deps/CMakeLists.txt:
```diff
@@ -177,8 +177,8 @@ set(UNIBILIUM_SHA256 29815283c654277ef77a3adcc8840db79ddbb20a0f0b0c8f648bd8cd49a
 
-set(LIBVTERM_URL https://www.leonerd.org.uk/code/libvterm/libvterm-0.1.4.tar.gz)
-set(LIBVTERM_SHA256 bc70349e95559c667672fc8c55b9527d9db9ada0fb80a3beda533418d782d3dd)
+set(LIBVTERM_URL http://www.leonerd.org.uk/code/libvterm/libvterm-0.3-RC1.tar.gz)
```
That should be visible on https

clason replied:
copy-pasta

leonerd commented on src/nvim/terminal.c:
```diff
@@ -887,7 +892,34 @@ static int term_settermprop(VTermProp prop, VTermValue *val, void *data)
+      term->title_size = MAX(frag.len, 1024);
+      term->title = xmalloc(sizeof(char *) * term->title_size);
+    } else if (term->title_len + frag.len > term->title_size) {
+        term->title_size *= 2;
+        term->title = xrealloc(term->title, sizeof(char *) * term->title_size);
```
I believe this security warning is upset about the inconsistent indentation level inside the `else if` block.

clason replied:
Yup, already fixed locally (just didn't want to interrupt and restart the CI run). Will push a fixed-up version shortly.

clason commented:
@cryptomilk I couldn't wait so I pushed the changes to your branch -- your patch and the libvterm-0.3-RC1 release passed all tests (I had to drop the VLA patch for MSVC, which is a nice bonus -- fixes #9540, finally).

Thank you (and @leonerd) for pushing this through! I will squash the commits and merge tonight or tomorrow unless someone objects. (@jamessan ?)

leonerd commented:
Well I *slightly* object, but only on the grounds you're using the -RC1 release candidate. That's supposed to be just a temporary "does it work?" test. I'm hoping in a few days if everyone says it's working fine, I'll cut a real 0.3 release and you can use that. But if you want to commit this for now on the expectation that you'll bump it soon then that's fine.

clason commented:
>  But if you want to commit this for now on the expectation that you'll bump it soon then that's fine.

Yes, exactly. This may help surface bugs that can be fixed for both your 0.3 and our 0.8 release.

clason commented:
(@carlocab also means that you can finally get rid of that resource in the Neovim formula -- to make up for the treesitter parsers ;))

carlocab commented:
> (@carlocab also means that you can finally get rid of that resource in the Neovim formula -- to make up for the treesitter parsers ;))

Amazing, now we have room for parsers.

Does this work with `libvterm` 0.2, though? I'm already predicting multiple issues opened about a broken `brew install --HEAD neovim`.

clason commented:
> Does this work with libvterm 0.2, though? I'm already predicting multiple issues opened about a broken brew install --HEAD neovim.

Any more than we're already getting because people have `brew installed libvterm` ;)

Yes, it builds and (from a quick test) works fine when building against 0.2 locally on macOS. (You won't get the nice reflow, of course...)

But I fully expect that by the time 0.8 is released, homebrew will have bottled libvterm-0.3, anyway.

carlocab commented:
Works for me: https://github.com/Homebrew/homebrew-core/pull/109791

carlocab commented:
I've merged Homebrew/homebrew-core#109791 if someone opens an issue about `brew install --HEAD neovim` being broken, tell them to `brew update` and try again. If that doesn't work, you can point them at the Homebrew [discussions page](https://github.com/Homebrew/discussions/discussions). (No issues about `--HEAD` installs, please.)

carlocab commented:
Oh, and for anyone who's looking for reflow with Homebrew `libvterm`, copy this patch:
```diff
diff --git a/Formula/libvterm.rb b/Formula/libvterm.rb
index 8f447b13e9b..ce0c391ac63 100644
--- a/Formula/libvterm.rb
+++ b/Formula/libvterm.rb
@@ -1,8 +1,9 @@
 class Libvterm < Formula
   desc "C99 library which implements a VT220 or xterm terminal emulator"
   homepage "http://www.leonerd.org.uk/code/libvterm/"
-  url "http://www.leonerd.org.uk/code/libvterm/libvterm-0.2.tar.gz"
-  sha256 "4c5150655438cfb8c57e7bd133041140857eb04defd0e544521c0e469258e105"
+  url "http://www.leonerd.org.uk/code/libvterm/libvterm-0.3-RC1.tar.gz"
+  version "0.3-RC1"
+  sha256 "441d1c372b84a0df12525100ab06c0366260fb4f6252abd1665ee4fa571b5134"
   license "MIT"
   version_scheme 1
 
```
Apply it with:
```sh
cd "$(brew --repository homebrew/core)"
pbpaste | patch -p1
brew upgrade --build-from-source libvterm
git restore -- Formula/libvterm.rb
```
